### ISSUE TITLE
Kubelet on Azure is making vmss calls even when useInstanceMetadata is set to true

### ISSUE THREAD

**Issue by matthias50** [kubernetes/kubernetes#93796]
**What happened**:

Kubelet is logging error messages every 10 seconds indicating failed vmss calls because of rate limiting. These calls prevent the subscription from recovering from whatever was the original cause of the rate limiting by making these API calls continuously.

**What you expected to happen**:
Kubelet does not make these API calls when useInstanceMetadata is set to true in the cloud config.

**How to reproduce it (as minimally and precisely as possible)**:
Start an azure node with kubelet command line like this:
```
/usr/bin/kubelet --cloud-provider=azure --cloud-config=/srv/kubernetes/kubelet/cloud.conf --hostname-override=dp-dev-3-az-eastus-etcd-1000002 --pod-infra-container-image=hcr.data.here.com/public_mirror/pause-amd64:3.1 --runtime-cgroups=/systemd/system.slice --kubeconfig=/srv/kubernetes/kubelet/kubeconfig --config=/srv/kubernetes/kubelet/kubelet-config.yaml --network-plugin=cni --cluster-dns=169.254.20.10 --node-labels=here.com/role=etcd,kubernetes.azure.com/managed=false,workload=default --non-masquerade-cidr=100.64.0.0/10 --register-schedulable=true --register-with-taints=node-role.kubernetes.io/master=:NoSchedule --container-runtime=remote --container-runtime-endpoint=/var/run/crio/crio.sock --v=2 --cni-bin-dir=/opt/cni/bin/ --cni-conf-dir=/etc/cni/net.d/ --pod-max-pids=2048 --feature-gates=SupportPodPidsLimit=true
```
Contents of /srv/kubernetes/kubelet/cloud.conf
```
{"tenantId": "redacted", "subscriptionId": "redacted", "location": "eastus", "resourceGroup": "dp-dev-3-az-eastus", "subnetName": "dp-dev-3-az-eastus-loadbalancer", "securityGroupName": "dp-dev-3-az-eastus-main-nsg", "loadBalancerName": "dp-dev-3-az-eastus-lb", "loadBalancerResourceGroup": "dp-dev-3-az-eastus", "cloud": "AzurePublicCloud", "vmType": "vmss", "vnetName": "dp-dev-3-az-eastus-vnet", "vnetResourceGroup": "dp-dev-3-az-eastus", "routeTableName": "dp-rd-az-eastus-cohesive-rt", "loadBalancerSku": "Standard", "maximumLoadBalancerRuleCount": 250, "excludeMasterFromStandardLB": true, "cloudProviderBackoffMode": "v2", "cloudProviderBackoff": true, "cloudProviderBackoffRetries": 6, "cloudProviderBackoffExponent": 1.5, "cloudProviderBackoffDuration": 5, "cloudProviderBackoffJitter": 1, "cloudProviderRatelimit": true, "cloudProviderRateLimitQPS": 10, "cloudProviderRateLimitBucket": 100, "cloudProviderRatelimitQPSWrite": 10, "cloudProviderRatelimitBucketWrite": 100, "useManagedIdentityExtension": true, "useInstanceMetadata": true, "providerKeyName": "k8s"}
```
After some unknown trigger occurs, kubelet will start logging errors like this:
```
Aug 07 14:50:39 dp-dev-3-az-eastus-etcd-1000002.internal.cloudapp.net kubelet[5254]: E0807 14:50:39.198941    5254 azure_vmss.go:569] VirtualMachineScaleSetsClient.List failed: compute.VirtualMachineScaleSetsClient#List: Failure responding to request: StatusCode=429 -- Original Error: autorest/azure: Service returned an error. Status=429 Code="OperationNotAllowed" Message="The server rejected the request because too many requests have been received for this subscription." Details=[{"code":"TooManyRequests","message":"{\"operationGroup\":\"HighCostGetVMScaleSet3Min\",\"startTime\":\"2020-08-07T14:50:39.1823549+00:00\",\"endTime\":\"2020-08-07T14:52:39.1823549+00:00\",\"allowedRequestCount\":280,\"measuredRequestCount\":937}","target":"HighCostGetVMScaleSet3Min"},{"code":"TooManyRequests","message":"{\"operationGroup\":\"HighCostGetVMScaleSet30Min\",\"startTime\":\"2020-08-07T14:50:39.1823549+00:00\",\"endTime\":\"2020-08-07T15:15:00+00:00\",\"allowedRequestCount\":1400,\"measuredRequestCount\":9229}","target":"HighCostGetVMScaleSet30Min"}] InnerError={"internalErrorCode":"TooManyRequestsReceived"}
Aug 07 14:50:39 dp-dev-3-az-eastus-etcd-1000002.internal.cloudapp.net kubelet[5254]: E0807 14:50:39.199063    5254 azure_vmss.go:711] error: ss.GetPrimaryInterface(dp-dev-3-az-eastus-etcd-1000002), ss.getVmssVM(dp-dev-3-az-eastus-etcd-1000002), err=compute.VirtualMachineScaleSetsClient#List: Failure responding to request: StatusCode=429 -- Original Error: autorest/azure: Service returned an error. Status=429 Code="OperationNotAllowed" Message="The server rejected the request because too many requests have been received for this subscription." Details=[{"code":"TooManyRequests","message":"{\"operationGroup\":\"HighCostGetVMScaleSet3Min\",\"startTime\":\"2020-08-07T14:50:39.1823549+00:00\",\"endTime\":\"2020-08-07T14:52:39.1823549+00:00\",\"allowedRequestCount\":280,\"measuredRequestCount\":937}","target":"HighCostGetVMScaleSet3Min"},{"code":"TooManyRequests","message":"{\"operationGroup\":\"HighCostGetVMScaleSet30Min\",\"startTime\":\"2020-08-07T14:50:39.1823549+00:00\",\"endTime\":\"2020-08-07T15:15:00+00:00\",\"allowedRequestCount\":1400,\"measuredRequestCount\":9229}","target":"HighCostGetVMScaleSet30Min"}] InnerError={"internalErrorCode":"TooManyRequestsReceived"}
Aug 07 14:50:39 dp-dev-3-az-eastus-etcd-1000002.internal.cloudapp.net kubelet[5254]: E0807 14:50:39.199088    5254 azure_vmss.go:418] error: ss.GetIPByNodeName(dp-dev-3-az-eastus-etcd-1000002), GetPrimaryInterface("dp-dev-3-az-eastus-etcd-1000002"), err=compute.VirtualMachineScaleSetsClient#List: Failure responding to request: StatusCode=429 -- Original Error: autorest/azure: Service returned an error. Status=429 Code="OperationNotAllowed" Message="The server rejected the request because too many requests have been received for this subscription." Details=[{"code":"TooManyRequests","message":"{\"operationGroup\":\"HighCostGetVMScaleSet3Min\",\"startTime\":\"2020-08-07T14:50:39.1823549+00:00\",\"endTime\":\"2020-08-07T14:52:39.1823549+00:00\",\"allowedRequestCount\":280,\"measuredRequestCount\":937}","target":"HighCostGetVMScaleSet3Min"},{"code":"TooManyRequests","message":"{\"operationGroup\":\"HighCostGetVMScaleSet30Min\",\"startTime\":\"2020-08-07T14:50:39.1823549+00:00\",\"endTime\":\"2020-08-07T15:15:00+00:00\",\"allowedRequestCount\":1400,\"measuredRequestCount\":9229}","target":"HighCostGetVMScaleSet30Min"}] InnerError={"internalErrorCode":"TooManyRequestsReceived"}
Aug 07 14:50:39 dp-dev-3-az-eastus-etcd-1000002.internal.cloudapp.net kubelet[5254]: I0807 14:50:39.199110    5254 azure_instances.go:58] NodeAddresses(dp-dev-3-az-eastus-etcd-1000002) abort backoff: compute.VirtualMachineScaleSetsClient#List: Failure responding to request: StatusCode=429 -- Original Error: autorest/azure: Service returned an error. Status=429 Code="OperationNotAllowed" Message="The server rejected the request because too many requests have been received for this subscription." Details=[{"code":"TooManyRequests","message":"{\"operationGroup\":\"HighCostGetVMScaleSet3Min\",\"startTime\":\"2020-08-07T14:50:39.1823549+00:00\",\"endTime\":\"2020-08-07T14:52:39.1823549+00:00\",\"allowedRequestCount\":280,\"measuredRequestCount\":937}","target":"HighCostGetVMScaleSet3Min"},{"code":"TooManyRequests","message":"{\"operationGroup\":\"HighCostGetVMScaleSet30Min\",\"startTime\":\"2020-08-07T14:50:39.1823549+00:00\",\"endTime\":\"2020-08-07T15:15:00+00:00\",\"allowedRequestCount\":1400,\"measuredRequestCount\":9229}","target":"HighCostGetVMScaleSet30Min"}] InnerError={"internalErrorCode":"TooManyRequestsReceived"}
Aug 07 14:50:39 dp-dev-3-az-eastus-etcd-1000002.internal.cloudapp.net kubelet[5254]: I0807 14:50:39.199137    5254 cloud_request_manager.go:115] Node addresses from cloud provider for node "dp-dev-3-az-eastus-etcd-1000002" not collected: compute.VirtualMachineScaleSetsClient#List: Failure responding to request: StatusCode=429 -- Original Error: autorest/azure: Service returned an error. Status=429 Code="OperationNotAllowed" Message="The server rejected the request because too many requests have been received for this subscription." Details=[{"code":"TooManyRequests","message":"{\"operationGroup\":\"HighCostGetVMScaleSet3Min\",\"startTime\":\"2020-08-07T14:50:39.1823549+00:00\",\"endTime\":\"2020-08-07T14:52:39.1823549+00:00\",\"allowedRequestCount\":280,\"measuredRequestCount\":937}","target":"HighCostGetVMScaleSet3Min"},{"code":"TooManyRequests","message":"{\"operationGroup\":\"HighCostGetVMScaleSet30Min\",\"startTime\":\"2020-08-07T14:50:39.1823549+00:00\",\"endTime\":\"2020-08-07T15:15:00+00:00\",\"allowedRequestCount\":1400,\"measuredRequestCount\":9229}","target":"HighCostGetVMScaleSet30Min"}] InnerError={"internalErrorCode":"TooManyRequestsReceived"}

Aug 07 14:50:49 dp-dev-3-az-eastus-etcd-1000002.internal.cloudapp.net kubelet[5254]: E0807 14:50:49.281364    5254 azure_vmss.go:569] VirtualMachineScaleSetsClient.List failed: compute.VirtualMachineScaleSetsClient#List: Failure responding to request: StatusCode=429 -- Original Error: autorest/azure: Service returned an error. Status=429 Code="OperationNotAllowed" Message="The server rejected the request because too many requests have been received for this subscription." Details=[{"code":"TooManyRequests","message":"{\"operationGroup\":\"HighCostGetVMScaleSet3Min\",\"startTime\":\"2020-08-07T14:50:49.2605019+00:00\",\"endTime\":\"2020-08-07T14:53:00+00:00\",\"allowedRequestCount\":280,\"measuredRequestCount\":998}","target":"HighCostGetVMScaleSet3Min"},{"code":"TooManyRequests","message":"{\"operationGroup\":\"HighCostGetVMScaleSet30Min\",\"startTime\":\"2020-08-07T14:50:49.2605019+00:00\",\"endTime\":\"2020-08-07T15:15:00+00:00\",\"allowedRequestCount\":1400,\"measuredRequestCount\":9290}","target":"HighCostGetVMScaleSet30Min"}] InnerError={"internalErrorCode":"TooManyRequestsReceived"}
Aug 07 14:50:49 dp-dev-3-az-eastus-etcd-1000002.internal.cloudapp.net kubelet[5254]: E0807 14:50:49.281541    5254 azure_vmss.go:711] error: ss.GetPrimaryInterface(dp-dev-3-az-eastus-etcd-1000002), ss.getVmssVM(dp-dev-3-az-eastus-etcd-1000002), err=compute.VirtualMachineScaleSetsClient#List: Failure responding to request: StatusCode=429 -- Original Error: autorest/azure: Service returned an error. Status=429 Code="OperationNotAllowed" Message="The server rejected the request because too many requests have been received for this subscription." Details=[{"code":"TooManyRequests","message":"{\"operationGroup\":\"HighCostGetVMScaleSet3Min\",\"startTime\":\"2020-08-07T14:50:49.2605019+00:00\",\"endTime\":\"2020-08-07T14:53:00+00:00\",\"allowedRequestCount\":280,\"measuredRequestCount\":998}","target":"HighCostGetVMScaleSet3Min"},{"code":"TooManyRequests","message":"{\"operationGroup\":\"HighCostGetVMScaleSet30Min\",\"startTime\":\"2020-08-07T14:50:49.2605019+00:00\",\"endTime\":\"2020-08-07T15:15:00+00:00\",\"allowedRequestCount\":1400,\"measuredRequestCount\":9290}","target":"HighCostGetVMScaleSet30Min"}] InnerError={"internalErrorCode":"TooManyRequestsReceived"}
Aug 07 14:50:49 dp-dev-3-az-eastus-etcd-1000002.internal.cloudapp.net kubelet[5254]: E0807 14:50:49.281601    5254 azure_vmss.go:418] error: ss.GetIPByNodeName(dp-dev-3-az-eastus-etcd-1000002), GetPrimaryInterface("dp-dev-3-az-eastus-etcd-1000002"), err=compute.VirtualMachineScaleSetsClient#List: Failure responding to request: StatusCode=429 -- Original Error: autorest/azure: Service returned an error. Status=429 Code="OperationNotAllowed" Message="The server rejected the request because too many requests have been received for this subscription." Details=[{"code":"TooManyRequests","message":"{\"operationGroup\":\"HighCostGetVMScaleSet3Min\",\"startTime\":\"2020-08-07T14:50:49.2605019+00:00\",\"endTime\":\"2020-08-07T14:53:00+00:00\",\"allowedRequestCount\":280,\"measuredRequestCount\":998}","target":"HighCostGetVMScaleSet3Min"},{"code":"TooManyRequests","message":"{\"operationGroup\":\"HighCostGetVMScaleSet30Min\",\"startTime\":\"2020-08-07T14:50:49.2605019+00:00\",\"endTime\":\"2020-08-07T15:15:00+00:00\",\"allowedRequestCount\":1400,\"measuredRequestCount\":9290}","target":"HighCostGetVMScaleSet30Min"}] InnerError={"internalErrorCode":"TooManyRequestsReceived"}
Aug 07 14:50:49 dp-dev-3-az-eastus-etcd-1000002.internal.cloudapp.net kubelet[5254]: I0807 14:50:49.281690    5254 azure_instances.go:58] NodeAddresses(dp-dev-3-az-eastus-etcd-1000002) abort backoff: compute.VirtualMachineScaleSetsClient#List: Failure responding to request: StatusCode=429 -- Original Error: autorest/azure: Service returned an error. Status=429 Code="OperationNotAllowed" Message="The server rejected the request because too many requests have been received for this subscription." Details=[{"code":"TooManyRequests","message":"{\"operationGroup\":\"HighCostGetVMScaleSet3Min\",\"startTime\":\"2020-08-07T14:50:49.2605019+00:00\",\"endTime\":\"2020-08-07T14:53:00+00:00\",\"allowedRequestCount\":280,\"measuredRequestCount\":998}","target":"HighCostGetVMScaleSet3Min"},{"code":"TooManyRequests","message":"{\"operationGroup\":\"HighCostGetVMScaleSet30Min\",\"startTime\":\"2020-08-07T14:50:49.2605019+00:00\",\"endTime\":\"2020-08-07T15:15:00+00:00\",\"allowedRequestCount\":1400,\"measuredRequestCount\":9290}","target":"HighCostGetVMScaleSet30Min"}] InnerError={"internalErrorCode":"TooManyRequestsReceived"}
Aug 07 14:50:49 dp-dev-3-az-eastus-etcd-1000002.internal.cloudapp.net kubelet[5254]: I0807 14:50:49.281747    5254 cloud_request_manager.go:115] Node addresses from cloud provider for node "dp-dev-3-az-eastus-etcd-1000002" not collected: compute.VirtualMachineScaleSetsClient#List: Failure responding to request: StatusCode=429 -- Original Error: autorest/azure: Service returned an error. Status=429 Code="OperationNotAllowed" Message="The server rejected the request because too many requests have been received for this subscription." Details=[{"code":"TooManyRequests","message":"{\"operationGroup\":\"HighCostGetVMScaleSet3Min\",\"startTime\":\"2020-08-07T14:50:49.2605019+00:00\",\"endTime\":\"2020-08-07T14:53:00+00:00\",\"allowedRequestCount\":280,\"measuredRequestCount\":998}","target":"HighCostGetVMScaleSet3Min"},{"code":"TooManyRequests","message":"{\"operationGroup\":\"HighCostGetVMScaleSet30Min\",\"startTime\":\"2020-08-07T14:50:49.2605019+00:00\",\"endTime\":\"2020-08-07T15:15:00+00:00\",\"allowedRequestCount\":1400,\"measuredRequestCount\":9290}","target":"HighCostGetVMScaleSet30Min"}] InnerError={"internalErrorCode":"TooManyRequestsReceived"}
```

Since useInstanceMetadata is set to true, kubelet should be using instance metadata instead of vmss calls in all cases. 
 
**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):

Client Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.9", GitCommit:"4fb7ed12476d57b8437ada90b4f93b17ffaeed99", GitTreeState:"clean", BuildDate:"2020-07-15T16:18:16Z", GoVersion:"go1.13.9", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.9", GitCommit:"4fb7ed12476d57b8437ada90b4f93b17ffaeed99", GitTreeState:"clean", BuildDate:"2020-07-15T16:10:45Z", GoVersion:"go1.13.9", Compiler:"gc", Platform:"linux/amd64"}

- Cloud provider or hardware configuration: Azure
- OS (e.g: `cat /etc/os-release`): 
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CENTOS_MANTISBT_PROJECT="CentOS-7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="7"

- Kernel (e.g. `uname -a`):
Linux dp-dev-3-az-eastus-etcd-1000000.internal.cloudapp.net 3.10.0-1127.13.1.el7.x86_64 #1 SMP Tue Jun 23 15:46:38 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
- Install tools: Custom terraform stack and scripts
- Network plugin and version (if this is a network-related bug):
- Others:

**Comment by andyzhangx** [comment]
@matthias50 could you check whether your node name is identical to name returned by instance meta data, I think it's not equal, so finally would invoke ARM API, related code is here: 
https://github.com/kubernetes/kubernetes/blob/v1.17.9/staging/src/k8s.io/legacy-cloud-providers/azure/azure_instances.go#L85-L98

you could ssh to one node, and run `curl -H Metadata:true "http://169.254.169.254/metadata/instance?api-version=2019-04-30" | jq` to check the instance meta data

**Comment by matthias50** [comment]
I can tell you without checking that it is not. We are overriding the node name with --hostname-override. We are doing this because, with our vmss scale set names, we were exceeding some DNS limitations.

**Comment by matthias50** [comment]
Here is an example:
Kubernetes Node name: dp-dev-3-az-eastus-worker-default-1000001
vmScaleSetName: dp-dev-3-az-eastus-worker-default-1-scale
hostname: dp-dev-3-az-eastus-worker-default-1000001.internal.cloudapp.net
curl -s -H Metadata:true "http://169.254.169.254/metadata/instance?api-version=2019-04-30" | jq -r '.compute.name': dp-dev-3-az-eastus-worker-default-1-scale_1

So, should the hostname and node name all equal `.compute.name` from the metadata?

**Comment by matthias50** [comment]
Ok, so I dug a bit further and found this line which looks suspicious: https://github.com/kubernetes/kubernetes/blob/v1.17.9/staging/src/k8s.io/legacy-cloud-providers/azure/azure_instances.go#L237. Shouldn't this take the value from hostname override if it is present?

**Comment by andyzhangx** [comment]
> dp-dev-3-az-eastus-worker-default-1000001.internal.cloudapp.net

hostname should be `dp-dev-3-az-eastus-worker-default-1000001` without `.internal.cloudapp.net` suffix

**Comment by matthias50** [comment]
> hostname should be `dp-dev-3-az-eastus-worker-default-1000001` without `.internal.cloudapp.net` suffix

Got it. This was how we were configuring things till recently. The root cause seems to be a race condition in our cloud-init code configuring the node. Our cloud-init scripts are disabling NetworkManager. In our particular CentOS 7 configuration, if you don't disable NetworkManager before it has a chance to run, it will change the network configuration to add the `.internal.cloudapp.net` suffix.

It would have made things *much* easier if kubelet would have had a fatal error in the case that useInstanceMetadata is set to true and the hostname != nodename. Would it be possible to make this change to prevent others from making the same mistake?

**Comment by andyzhangx** [comment]
> > hostname should be `dp-dev-3-az-eastus-worker-default-1000001` without `.internal.cloudapp.net` suffix
> 
> Got it. This was how we were configuring things till recently. The root cause seems to be a race condition in our cloud-init code configuring the node. Our cloud-init scripts are disabling NetworkManager. In our particular CentOS 7 configuration, if you don't disable NetworkManager before it has a chance to run, it will change the network configuration to add the `.internal.cloudapp.net` suffix.
> 
> It would have made things _much_ easier if kubelet would have had a fatal error in the case that useInstanceMetadata is set to true and the hostname != nodename. Would it be possible to make this change to prevent others from making the same mistake?

I am thinking write psudo code like this to make it work:
`return (metadataVMName == nodeName), err` -->

```
segments := strings.Split(nodeName, ".")
hostName := segments[0]
return metadataVMName == hostName, err
```

**Comment by matthias50** [comment]
> I am thinking write psudo code like this to make it work:
> `return (metadataVMName == nodeName), err` -->
> 
> ```
> segments := strings.Split(nodeName, ".")
> hostName := segments[0]
> return metadataVMName == hostName, err
> ```

This would do the right thing in the precise case that the `hostname` has a domain suffix and hostname_overrride is being used to tell kubelet to use only the portion before the dot as the node name. In other cases, it would be counter-productive. For instance, suppose I had the hostname dp-dev-3-az-eastus-worker-default-1000001.internal.cloudapp.net and then matched the nodename to also be dp-dev-3-az-eastus-worker-default-1000001.internal.cloudapp.net by not passing hostname-override. In that case, it looks like the proposed fix would actually create a problem.

`isCurrentInstance` should, by definition always be true when it is called in the context of kubelet because kubelet only operates on itself. So I think a more general fix would be to compare `metadataVMName == nodeName` and if this condition is false and the context of the comparison is kubelet and not the controller-manager, then return some error. Alternatively, the calling function `NodeAddresses` could perform a similar check before operating on the assumption that the node is non-local here: https://github.com/kubernetes/kubernetes/blob/4fb7ed12476d57b8437ada90b4f93b17ffaeed99/staging/src/k8s.io/legacy-cloud-providers/azure/azure_instances.go#L91

**Comment by andyzhangx** [comment]
> > I am thinking write psudo code like this to make it work:
> > `return (metadataVMName == nodeName), err` -->
> > ```
> > segments := strings.Split(nodeName, ".")
> > hostName := segments[0]
> > return metadataVMName == hostName, err
> > ```
> 
> This would do the right thing in the precise case that the `hostname` has a domain suffix and hostname_overrride is being used to tell kubelet to use only the portion before the dot as the node name. In other cases, it would be counter-productive. For instance, suppose I had the hostname dp-dev-3-az-eastus-worker-default-1000001.internal.cloudapp.net and then matched the nodename to also be dp-dev-3-az-eastus-worker-default-1000001.internal.cloudapp.net by not passing hostname-override. In that case, it looks like the proposed fix would actually create a problem.
> 
> `isCurrentInstance` should, by definition always be true when it is called in the context of kubelet because kubelet only operates on itself. So I think a more general fix would be to compare `metadataVMName == nodeName` and if this condition is false and the context of the comparison is kubelet and not the controller-manager, then return some error. Alternatively, the calling function `NodeAddresses` could perform a similar check before operating on the assumption that the node is non-local here:
> 
> https://github.com/kubernetes/kubernetes/blob/4fb7ed12476d57b8437ada90b4f93b17ffaeed99/staging/src/k8s.io/legacy-cloud-providers/azure/azure_instances.go#L91

that would change the behavior of current `InstanceID` func logic: 
  if not found in UseInstanceMetadata, then get addresses from Azure ARM API.
I think use Azure ARM API is still valid in some case.

what about such fix:
```
segments := strings.Split(nodeName, ".")
if len(segments) > 0 && metadataVMName ==  segments[0] {
  return true, err
}
return metadataVMName == nodeName , err
```

**Comment by matthias50** [comment]
> I think use Azure ARM API is still valid in some case.

If execution context is kubelet and useInstanceMetadata is true, I can't think of any case where ARM API is valid. Am I missing something?

That said, I think your proposed fix would work. It is just a bit subtle and my concern is that I'm missing something and we would end up having ARM calls from kubelet in some scenario I'm not thinking of.

**Comment by andyzhangx** [comment]
+ @feiskyer for comments

**Comment by feiskyer** [comment]
Is the changed VM name reflected on the VMSS instance data model (e.g. GET VMSS instance body)?

**Comment by matthias50** [comment]
> Is the changed VM name reflected on the VMSS instance data model (e.g. GET VMSS instance body)?

No, from the example above, `curl -s -H Metadata:true "http://169.254.169.254/metadata/instance?api-version=2019-04-30" | jq -r '.compute.name'` doesn't change and is `dev-3-az-eastus-worker-default-1-scale_1`

**Comment by feiskyer** [comment]
@matthias50 then I think the call is expected because current implementations depend on either IMDS or Azure API to determine the node metadata. If the information is not in both of them, then it's hard to match the NodeName to VM. Are you able to use the auto-generated VM Name from VMSS?

**Comment by matthias50** [comment]
To be clear, we have solved the issue by ensuring the `hostname` is the codename. Disabling NetworkManager does this. My remaining concern is that when kubelet runs this code and it finds that hostname != nodename, and  useInstanceMetadata is true that this should be a fatal error. I say that because it is impossible for kubelet to actually use instance metadata in this case. If kubelet had behaved in this fashion, it would have saved our team over a week of lost productivity.

**Comment by feiskyer** [comment]
@matthias50 Glad to hear the issue has been resolved. Yep, we have assumption that hostname should be the same as nodename. We added the fallback to Azure ARM API because the same function may be invoked by two component: kubelet and cloud-controller-manager. Since cloud-controller-manager is running on master nodes, it needs to invoke Azure API for getting other VMs metadata. We would resolve this issue with out-of-tree cloud provider component cloud-node-manager. 

See the out-of-tree cloud provider here: https://github.com/kubernetes-sigs/cloud-provider-azure (it is not GA yet, hence not recommended for production usage).
